crates/block-producer/src/mempool/mod.rs

```diff
@@ -73,32 +77,161 @@ impl BlockNumber {
     }
 }
 
+// MEMPOOL BUDGET
+// ================================================================================================
+
+/// Limits placed on a batch's contents.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct BatchBudget {
+    /// Maximum number of transactions allowed in a batch.
+    pub transactions: usize,
+    /// Maximum number of input notes allowed.
+    pub input_notes: usize,
+    /// Maximum number of output notes allowed.
+    pub output_notes: usize,
+    /// Maximum number of updated accounts.
+    pub accounts: usize,
+}
+
+/// Limits placed on a blocks's contents.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct BlockBudget {
+    /// Maximum number of batches allowed in a block.
+    pub batches: usize,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum BudgetStatus {
+    Depleted,
+    Undepleted,
+}
+
+impl Default for BatchBudget {
+    fn default() -> Self {
+        Self {
+            transactions: SERVER_MAX_TXS_PER_BATCH,
+            input_notes: MAX_INPUT_NOTES_PER_BATCH,
+            output_notes: MAX_OUTPUT_NOTES_PER_BATCH,
+            accounts: MAX_ACCOUNTS_PER_BATCH,
+        }
+    }
+}
+
+impl Default for BlockBudget {
+    fn default() -> Self {
+        Self { batches: SERVER_MAX_BATCHES_PER_BLOCK }
+    }
+}
+
+impl BatchBudget {
+    /// Attempts to consume the transaction's resources from the budget.
+    ///
+    /// Returns [BudgetStatus::Depleted] if the transaction would exceed the remaining budget,
+    /// otherwise returns [BudgetStatus::Undepleted]
+    #[must_use]
+    fn check_then_deplete(&mut self, tx: &AuthenticatedTransaction) -> BudgetStatus {
+        // This type assertion reminds us to update the account check if we ever support multiple
+        // account updates per tx.
+        let _: miden_objects::accounts::AccountId = tx.account_update().account_id();
+        const ACCOUNT_COUNT: usize = 1;
```

Comment:
I think, we should rename this constant in order to have clearer name. Something like `TX_MIN_ACCOUNT_COUNT`.